android/src/main/java/com/pagopa/ioreactnativeintegrity/IoReactNativeIntegrityModule.kt

```diff
@@ -1,25 +1,315 @@
 package com.pagopa.ioreactnativeintegrity
 
+import android.content.pm.PackageManager
+import android.os.Build
+import android.security.keystore.KeyGenParameterSpec
+import android.security.keystore.KeyInfo
+import android.security.keystore.KeyProperties
+import android.security.keystore.KeyProperties.SECURITY_LEVEL_STRONGBOX
+import android.security.keystore.KeyProperties.SECURITY_LEVEL_TRUSTED_ENVIRONMENT
+import android.security.keystore.KeyProperties.SECURITY_LEVEL_UNKNOWN_SECURE
+import android.util.Base64
+import androidx.annotation.RequiresApi
+import com.facebook.react.bridge.Promise
 import com.facebook.react.bridge.ReactApplicationContext
 import com.facebook.react.bridge.ReactContextBaseJavaModule
 import com.facebook.react.bridge.ReactMethod
-import com.facebook.react.bridge.Promise
+import com.facebook.react.bridge.WritableMap
+import com.facebook.react.bridge.WritableNativeMap
+import com.google.android.gms.common.GoogleApiAvailability
+import com.google.android.gms.tasks.Task
+import com.google.android.play.core.integrity.IntegrityManagerFactory
+import com.google.android.play.core.integrity.StandardIntegrityManager
+import com.google.android.play.core.integrity.StandardIntegrityManager.StandardIntegrityToken
+import com.google.android.play.core.integrity.StandardIntegrityManager.StandardIntegrityTokenProvider
+import com.google.android.play.core.integrity.StandardIntegrityManager.StandardIntegrityTokenRequest
+import java.security.KeyFactory
+import java.security.KeyPair
+import java.security.KeyPairGenerator
+import java.security.KeyStore
+import java.security.PrivateKey
+import java.security.spec.ECGenParameterSpec
 
+/**
+ * React Native bridge for Google Play Integrity API and key attestation.
+ * @property integrityTokenProvider the token integrity provider which should be initialized by calling [prepareIntegrityToken]
+ * @property keyStore the key store instance used to generate hardware backed keys and get a key attestation via [getAttestation]
+ */
 class IoReactNativeIntegrityModule(reactContext: ReactApplicationContext) :
   ReactContextBaseJavaModule(reactContext) {
 
+  private lateinit var integrityTokenProvider: StandardIntegrityTokenProvider
+
+  private val keyStore = KeyStore.getInstance(KEYSTORE_PROVIDER)
+
+  /**
+   * Constructor which initializes the keystore engine.
+   */
+  init {
+    keyStore.load(null)
```

Comment:
As this is the 'init' function, how can we inform the React Native app that this initialization process has failed? Should we consider relocating the load operation closer to its usage?

Reply:
In io-react-native-crypto we did this:

```
private val keyStore: KeyStore? by lazy {
    try {
      KeyStore.getInstance(KEYSTORE_PROVIDER).also {
        it.load(null)
      }
    } catch (e: Exception) {
      null
    }
  }
```

This try to initialize the `keyStore` at the first use and return null otherwise. Then at every usage of the `keyStore` we leverage the Kotlin null safety feature to handle the `null` case. What do you think?

Reply:
It makes sense to me, applied in 77967ffc11c1180a96c2a875d78f7cc944efe987, thanks for the suggestion.

Reply:
Perfect. So we handle the scenario were we have an empty array

`var attestations = arrayOf<String>()`

in case of keyStore loading failure, the same way if we have a loading success but no attestations. Is this a good handling or we prefer something like:

```
keyStore?.let {
          val chain = it.getCertificateChain(alias)
          // The certificate chain consists of an array of certificates, thus we concat them into a string
          var attestations = arrayOf<String>()
          chain.forEachIndexed { _, certificate ->
            val cert = Base64.encodeToString(certificate.encoded, Base64.DEFAULT)
            attestations += cert
          }
          val concatenatedAttestations = attestations.joinToString(",")
          val encodedAttestation =
            Base64.encodeToString(concatenatedAttestations.toByteArray(), Base64.DEFAULT)
          promise.resolve(encodedAttestation)
        } ?: ModuleException.KEYSTORE_NOT_INITIALIZED.reject(
          promise
        ) 
```

Reply:
I see, what I did is actually wrong as an empty array is a valid case. Fixed in d42b80245072ce0e61923798b8a68f47693c86b6, thanks.